README
	slsnif - Serial line sniffer



  Copyright (C) 2001 Yan Gurtovoy

  slsnif operates by opening a pty and linking it to the serial port.
  It takes following parameters / options:

  1. Serial port to open (required).
  2. Name of the file to direct output to (optional, defaults to stdout).
  3. Desired baudrate (optional, defaults to 9600 baud).
  4. Timestamp On/Off (optional, defaults to Off).
  5. Print ascii values in hex On/Off (optional, defaults to Off)
  6. Print number of bytes transmitted On/Off (optional, defaults to Off).
  7. Optional colors for timestamp, number of bytes transmitted,
     and normal output.
  8. Lock port On/Off (optional, defaults to On).
  9. Use Unix98 ptys instead of BSD ptys (optional, defaults to BSD style).
  10. Second serial port to open. If specified, this port will be used
     instead of a pty, thus providing an ability to log data between
     two serial ports.
  11. Filenames for dumping raw data into, separate files are used for
      input (from device) and output (from host) data.

  See man page for details on syntax, here are a couple of examples...
  
  Examples:
    slsnif /dev/ttyS1
        -- opens port /dev/ttyS1, sets speed to 9600 baud, output is sent
           to stdout
    slsnif /dev/ttyS0 -l log.txt
        -- opens port /dev/ttyS0, sets speed to 9600 baud, output is sent
           to file log.txt
    slsnif /dev/ttyS2 --log log1.txt -s 2400
        -- opens port /dev/ttyS2, sets speed to 2400 baud, output is sent
           to file log1.txt
    slsnif /dev/ttyS1 --speed 4800
        -- opens port /dev/ttyS1, sets speed to 4800 baud, output is sent
           to stdout.
    slsnif -p /dev/ttyS0 /dev/ttyS1
        -- opens ports /dev/ttyS0 and /dev/ttyS1. No pty is opened
	   in this case. Data coming from /dev/ttyS1 is considered
	   to be coming from a device and data coming from /dev/ttyS0
	   is considered to be coming from host. This consideration
	   is purely cosmetic, it only affects the way output works,
	   i.e. which chunks of data will be placed on lines that
	   begin with 'Device' and which chunks will go on lines
	   that begin with 'Host'. If pty is used, data from pty is
	   considered to be 'from host' and data from port is 'from
	   device'.
    slsnif -h
        -- displays help.           

  RC file:
    slsnif can read some of options from a configuration file
    '.slsnifrc', located in user's home directory. The following
    options can be specified in this file:
    
      RC-file option        corresponding command line option
      ------------------    ---------------------------------
      TOTALBYTES ON/OFF     -b (--bytes)
      TIMESTAMP  ON/OFF     -t (--timestamp)
      DISPLAYHEX ON/OFF     -x (--hex)
      NOLOCK     ON/OFF     -n (--nolock)
      SYSVPTY    ON/OFF     -u (--unix98)
      COLOR      <color>    --color <color>
      TIMECOLOR  <color>    --timecolor <color>
      BYTESCOLOR <color>    --bytescolor <color>

    Valid values for <color> are:
      black, red, green, yellow, blue, magenta, cyan, white,
      brightblack, brightred, brightgreen, brightyellow,
      brightblue, brightmagenta, brightcyan, and brightwhite.
            
    Options provided on the command line always override the
    ones read from the rc-file.
    
    An example of rc-file with comments (slsnifrc-example) is
    included in the distribution.
   
  How to use it:
    - start slsnif on a given port, note pty it has opened. The pty name is
      also available from file /tmp/slsnif_pty (useful for scripting).
    - configure controlling software for the device being debugged to use this
      pty instead of the real port, start controlling software.
    - while slsnif is running you can send it a SIGHUP signal, which would cause the
      rotation of all log files (i.e. file pointers will be moved to the beginning of
      the file).
    - while slsnif is running you can also send it a SIGUSR1 signal, which would cause
      immediate resynchronization of the port settings.
    - when you're done, stop slsnif by hitting <Ctrl + C>.
      
  slsnif is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

Comments are welcome.

	- Yan Gurtovoy <ymg@dakotacom.net>


ChangeLog
03/13/2001
	* 0.2.7:
	    -	first public release.

04/09/2001
        * 0.2.8:
	    -	rewrote command line parameters handling.
	    -	added a couple of output messages.

04/24/2001
	* 0.3.0:
	    -	added option to timestamp output
	    -	added option to print number of bytes transmitted

04/25/2001
	* 0.3.1:
	    -  fixed the infinite loop when looking for available pty bug
		(those typos will kill me one day :)

06/09/2001
	* 0.3.2:
	    -  added optional colors for terminal output.
	    -  fixed a bug with logfile name including '=' when
	       specified as `--log=<filename>`
	       
09/25/2001
	* 0.3.3:
	    -  fixed a bug with shell piping and output redirection
	       (`slsnif <parameters> | more` and Co. should work fine now).

10/21/2001
	* 0.3.4:
	    -  now compiles on a Sun SPARC Solaris 7 machine, thanks
	       to Heiko Nardmann (h.nardmann@secunet.de)

12/12/2001
	* 0.4.0:
	    -  added ability to lock the port ('-n' and '--nolock'
	       parameters).
	    -  added ability to use a real serial port instead of a
	       pty, thus enabling data logging between two ports.
	       ('-p' and '--port2' parameters).
	    -  now port parameters (termios structures) of both ports
	       are synchronized as soon as first chunk of data appears
	       on a host (pty) side, so setting baudrate is necessary
	       only if controlling software doesn't set it.
	    -  rewrote timestamping code to give more precise output
	       if ftime() function is present. New code is a bit slower
	       than the old one, if that causes problems (unlikely) one
	       can always force slsnif to use old code by changing line
	           #define HAVE_SYS_TIMEB_H  1
	       to
	           #undef HAVE_SYS_TIMEB_H
	       in config.h file before compilation.
	    -  added ability to read certain parameters from rc-file,
	       see README and man page for details on syntax.
	       An example of such file (slsnifrc-example) is included
	       in the distribution.
	    -  fixed some minor bugs. 

04/21/2002
	* 0.4.1:
	    -  Thanks to Jan Kratochvil (short@ucw.cz), slsnif is now able
	       to dump raw data into file(s) ('-i' and '--in-tee' parameters
	       for data from device, '-o' and '--out-tee' parameters for data
	       from host). Multiple '-i' and '-o' arguments are allowed if
	       more than one copy of raw data is necessary.
	       
03/31/2003
	* 0.4.2:
	    -  Added ability to move logfile pointer back to the beginning of the
	       file upon cathing the HUP signal for using with logrotate.
	    -  Added some extra error checking

06/16/2004
	* 0.4.3:
	    -  Increased possible length of the port name to 128 chars
	       (by requests of those people with non-standard device names)
            -  Added ability to display ascii values in hex ('-i' and '--hex'
               parameters). This can also be turned on/off from the rc-file.
            -  Fixed a bug that caused slsnif to lose first couple of chars
               when writing data to log file(s).
            -  slsnif now writes the name of opened pty to a file /tmp/slsnif_pty
               as well as to STDOUT. If it's unable to open this file, an error
               message is logged, but execution continues normally.

02/05/2005
        * 0.4.4:
            -  Added ability to resynchronize ports at any time by sending a
               SIGUSR1 signal to slsnif. Useful for debugging applications
               that change port parameters mid-way, after starting the
               transmission. One example is pppd daemon, which sets initial
               parameters in the chat script, and changes them later, after
               chat script terminates.
            -  Added ability to use SYSV (Unix98) ptys instead of BSD ones
               (see '-u' and '--unix98' parameters). This option can also be
               turned on/off from rc-file.


src/slsnif.h

/*  slsnif.h
 *  Copyright (C) 2001 Yan Gurtovoy (ymg@dakotacom.net)
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

#include <stdlib.h>
#include <stdio.h>
#include <ctype.h>
#include <unistd.h>
#include <fcntl.h>
#include <errno.h>
#include <signal.h>
#include <wait.h>
#include <sys/stat.h>
#include <sys/types.h>
#ifdef HAVE_GETOPT_H
#include <getopt.h>
#endif
#ifdef HAVE_SYS_TIMEB_H
#include <sys/timeb.h>
#endif
#ifdef HAVE_TIME_H
#include <time.h>
#endif
#include "common.h"

#define OPTSTR      "bnhtuxp:l:s:i:o:" /* list of short options for getopt_long() */
#define PORT_OUT    "\nHost   --> "
#define PORT_IN     "\nDevice --> "
#define DEFPTRNAME  "/dev/pty??"
#define TMPPATH     "/tmp/slsnif_pty"
#define TOTALBYTES  "Total bytes transmitted:"
#define MEMFAIL     "Memory allocation failed"
#define DIFFFAIL    "First and second ports can not be the same device"
#define PTYFAIL     "Failed to open a pty"
#define PORTFAIL    "Failed to open port"
#define LOGFAIL     "Failed to open log file, defaulting to stdout"
#define TMPFAIL     "Failed to open temp. file to save opened pty name, continuing..."
#define SELFAIL     "`Select` failed"
#define RPORTFAIL   "Error reading from port"
#define RPTYFAIL    "Error reading from pty"
#define CLOSEFAIL   "Failed to close file - data loss may occur"
#define REMFAIL     "Failed to delete file"
#define RAWFAIL     "Failed to switch into raw mode"
#define SYNCFAIL    "Failed to synchronize ttys"
#define PIPEFAIL    "Failed to create pipe"
#define FORKFAIL    "Failed to create child process"
#define RPIPEFAIL   "Error reading from pipe"
#define TEEFAIL     "Failed to create tee file"
#define TEEWRTFAIL  "Failed to write to tee file"

#define max(x,y)	((x) > (y) ? (x) : (y))

#define BUFFSIZE    32
#define PRFXSIZE    12
#define DEFBAUDRATE B9600
#define WHITE       15
#define PORTNAMELEN 128

static tty_struct   tty_data;
static pid_t        pid = -1;
static int          reseek = FALSE;
static int          needsync = TRUE;

/* baudrates */
typedef struct _spd_struct {
    speed_t speed;
    char    *spdstr;
} spd_struct;

typedef struct _tee_entry {
     struct _tee_entry *next;
     char   *name;
     int    fd;
} tee_entry;

/*
 * tee_files[0] - linked list of raw input files
 * tee_files[1] - linked list of raw output files
 * entry, entryp, tmp_entry - index variables
 */
static tee_entry *tee_files[2], *tmp_entry, *entry, **entryp;

static spd_struct baudrates[] = {
    {B50,         "50"}, {B75,         "75"}, {B110,         "110"},
    {B134,       "134"}, {B150,       "150"}, {B200,         "200"},
    {B300,       "300"}, {B600,       "600"}, {B1200,       "1200"},
    {B1800,     "1800"}, {B2400,     "2400"}, {B4800,       "4800"},
    {B9600,     "9600"}, {B19200,   "19200"}, {B38400,     "38400"},
    {B57600,   "57600"}, {B115200, "115200"}, {0,             NULL}
};

static clr_struct colors[] = {
    {"black",         "\033[0;30m"},
    {"red",           "\033[0;31m"},
    {"green",         "\033[0;32m"},
    {"yellow",        "\033[0;33m"},
    {"blue",          "\033[0;34m"},
    {"magenta",       "\033[0;35m"},
    {"cyan",          "\033[0;36m"},
    {"white",         "\033[0;37m"},
    {"brightblack",   "\033[1;30m"},
    {"brightred",     "\033[1;31m"},
    {"brightgreen",   "\033[1;32m"},
    {"brightyellow",  "\033[1;33m"},
    {"brightblue",    "\033[1;34m"},
    {"brightmagenta", "\033[1;35m"},
    {"brightcyan",    "\033[1;36m"},
    {"brightwhite",   "\033[1;37m"},
    {NULL,            NULL        }
};

/* function prototypes */
pid_t dev_getlock(char *dev);
pid_t dev_setlock(char *dev);
int dev_unlock(char *dev);
void readRC(tty_struct *ptr);
int getpt(void);
int grantpt(int fd);
char *ptsname(int fd);
int unlockpt(int fd);
